python/pyarrow/lib_gpu.pyx

```diff
@@ -0,0 +1,809 @@
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+import six
+from pyarrow.compat import tobytes
+from pyarrow.lib cimport *
+from pyarrow.includes.libarrow_gpu cimport *
+from pyarrow.lib import py_buffer, allocate_buffer
+cimport cpython as cp
+
+cdef class CudaDeviceManager:
+    """ CUDA device manager.
+    """
+
+    def __cinit__(self):
+        check_status(CCudaDeviceManager.GetInstance(& self.manager))
+
+    @property
+    def num_devices(self):
+        """ Return the number of GPU devices.
+        """
+        return self.manager.num_devices()
+
+    def get_context(self, gpu_number=0):
+        """Get the shared CUDA driver context for a particular device.
+
+        Parameters
+        ----------
+        gpu_number : int
+          Specify the gpu device for which the CUDA driver context is
+          requested.
+
+        Returns
+        -------
+        ctx : CudaContext
+          CUDA driver context.
+
+        """
+        cdef shared_ptr[CCudaContext] ctx
+        check_status(self.manager.GetContext(gpu_number, & ctx))
+        return pyarrow_wrap_cudacontext(ctx)
+
+    def create_new_context(self, gpu_number):
+        """ Create a new context for a given device number.
+
+        Use get_context for general code.
+        """
+        cdef shared_ptr[CCudaContext] ctx
+        check_status(self.manager.CreateNewContext(gpu_number, & ctx))
+        return pyarrow_wrap_cudacontext(ctx)
+
+    def allocate_host(self, nbytes):
+        """ Allocate host memory.
+
+        Parameters
+        ----------
+        nbytes : int
+          Specify the number of bytes to be allocated.
+
+        Returns
+        -------
+        hbuf : CudaHostBuffer
+          Buffer of allocated host memory.
+        """
+        cdef shared_ptr[CCudaHostBuffer] buf
+        check_status(self.manager.AllocateHost(nbytes, & buf))
+        return pyarrow_wrap_cudahostbuffer(buf)
+
+    def free_host(self, hbuf):
+        """ Free host memory allocated via allocate_host.
+
+        Parameters
+        ----------
+        hbuf : CudaHostBuffer
+          Specify host buffer to be deallocated.
+        """
+        if not pyarrow_is_cudahostbuffer(hbuf):
+            raise TypeError('expected CudaHostBuffer instance')
+        cdef CudaHostBuffer buf = < CudaHostBuffer > (hbuf)
+        if not buf._freed:
+            check_status(self.manager.FreeHost(
+                buf.host_buffer.get().mutable_data(), hbuf.size))
+            buf._freed = True
```

Comment:
Do we want or need to expose this API?

Reply:
I don't think so, because it doesn't handle automatic buffer management (`CudaHostBuffer::~CudaHostBuffer` crashes).

Reply:
Usage of FreeHost is eliminated from pyarrow.